src/pdn/test/pdn_aux.py

```diff
@@ -0,0 +1,676 @@
+
+
+valid_orientations = ["R0", "R90", "R180", "R270", "MX", "MY", "MXR90", "MYR90"]
+lef_orientations = {"N": "R0", "FN": "MY", "S": "R180", "FS": "MX", "E": "R270",
+                        "FE": "MYR90", "W": "R90", "FW": "MXR90"}
+
+def get_orientations(orientations):
+    if orientations == []:
+        return []
+    checked_orientations = []
+    for orient in orientations:
+      if orient in valid_orientations:
+        checked_orientations.append(orient)
+      elif orient in lef_orientations:
+        checked_orientations.append(lef_orientations[orient])
+      else:
+        vld = " ".join(valid_orientations)
+        utl.error(utl.PDN, 1036, f"Invalid orientation {orient} specified, " +
+                                 f"must be one of {vld}")
+
+    return checked_orientations
+
+
+def add_pdn_connect(design, *,
+                    grid="",
+                    layers=None,
+                    cut_pitch=[0, 0],
+                    fixed_vias=[],
+                    max_rows=0,
+                    max_columns=0,
+                    ongrid=[],       # list of layer names that should be on grid?
+                    split_cuts={},   # dictionary of layer name to pitch
+                    dont_use_vias=""):
+    pdngen = design.getPdnGen()
+    check_design_state(design, "add_pdn_connect")
+
+    if layers is None:
+        utl.error(utl.PDN, 1537, "The 'layers' argument is required.")
+    elif len(layers) != 2:
+        utl.error(utl.PDN, 1538, "The layers argument must contain two layers.")
+
+    l0 = get_layer(design, layers[0])
+    l1 = get_layer(design, layers[1])
+
+    if len(cut_pitch) != 2:
+        utl.error(utl.PDN, 1539, "The cut_pitch argument must contain two enties.")
+                   
+    cut_pitch = [design.micronToDBU(l) for l in cut_pitch]
+
+    fixed_generate_vias = []
+    fixed_tech_vias = []
+    for via in fixed_vias:
+        tech_via = design.getTech().getDB().getTech().findVia(via)
+        generate_via = design.getTech().getDB().getTech().findViaGenerateRule(via)
+        if tech_via is None and generate_via is None:
+            utl.error(utl.PDN, 1540, f"Unable to find via: {via}")
+        if bool(tech_via):
+            fixed_tech_vias.append(tech_via)
+        if bool(generate_via):
+            fixed_generate_vias.append(generate_via)
+
+    ongrid_list = [get_layer(l) for l in ongrid]
+
+    split_cuts_layers  = [get_layer(l) for l in split_cuts.keys()]
+    split_cuts_pitches = [design.micronToDBU(x) for x in split_cuts.values()]
+    split_cuts_dict = dict(zip(split_cuts_layers, split_cuts_pitches))
+    
+    for g in pdngen.findGrid(grid):
+        pdngen.makeConnect(g, l0, l1, cut_pitch[0], cut_pitch[1],
+                           fixed_generate_vias, fixed_tech_vias,
+                           max_rows, max_columns, ongrid_list,
+                           split_cuts_dict, dont_use_vias)
+
+
+def pdngen_db(design, failed_via_report="", skip_trim=False, dont_add_pins=False):
```

Comment:
I'm not sure what the thought is on this, but it seems like these functions should either get exposed to the users or we need to go through the tcl and python setups to move a lot of the error checking here moved into C

Reply:
pdngen_db is just the Python version of the tcl proc pdngen.  I just added the _db to avoid name clash / shadowing issues with the name pdngen. Moving the error checking into C++ is a good goal, but that is going to have to be done in stages.